include/kelcoro/thread_pool.hpp

```diff
@@ -81,8 +106,45 @@ struct task_queue {
     assert(!!nodes);
     return nodes;
   }
+
+  // precondition: node && node->task && node.status == ok
+  // executor interface
+  void attach(task_node* node) noexcept {
+    assert(node && node->task && node->status == schedule_errc::ok);
+    push(node);
+  }
 };
 
+static_assert(executor<task_queue>);
+
+namespace noexport {
+
+// We push the deadpill, and behind it ourselves,
+// in a chain, so that the deadpill is definitely alive.
+struct deadpill_pusher {
+  task_queue& queue;
+  task_node this_node;
+  task_node pill = task_node::deadpill();
+
+  static bool await_ready() noexcept {
+    return false;
+  }
+  void await_suspend(std::coroutine_handle<> handle) {
+    this_node.task = handle;
+    pill.next = &this_node;
+    queue.push_list(&pill);
+  }
+  void await_resume() noexcept {
+    assert(this_node.status == schedule_errc::cancelled);
+  }
+};
+
+inline dd::job push_deadpill(task_queue& queue) {
```

Comment:
i think it would be nice to have
```cpp
// produces all tasks from queue and automatically handles deadpills, task cancelling etc
dd::generator<coroutine_handle<>> tasks_of(task_queue& queue KELCORO_LIFETIMEBOUND);
```